File: SDK/CSharp/EdgeLinkUdpClient.cs
using System;
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace EdgeLink
{
    /// <summary>UDP receiver — binds to a local port and fires OnMessage for each packet.</summary>
    public class EdgeLinkUdpClient : IDisposable
    {
        public event Action<string>?    OnMessage;
        public event Action<Exception>? OnError;

        public int  LocalPort { get; }
        public bool IsRunning => !disposed && !cts.IsCancellationRequested;

        private UdpClient?              udp;
        private CancellationTokenSource cts = new();
        private readonly ConcurrentQueue<string> queue = new();
        private bool disposed;

        public EdgeLinkUdpClient(int localPort)
        {
            LocalPort = localPort;
        }

        public void Start()
        {
            if (disposed) throw new ObjectDisposedException(nameof(EdgeLinkUdpClient));
            cts = new CancellationTokenSource();
            udp = new UdpClient(LocalPort);
            _ = Task.Run(() => ReceiveLoopAsync(cts.Token), cts.Token);
        }

        private async Task ReceiveLoopAsync(CancellationToken ct)
        {
            while (!ct.IsCancellationRequested)
            {
                try
                {
                    var result = await udp!.ReceiveAsync(ct);
                    string msg = Encoding.UTF8.GetString(result.Buffer).Trim();
                    if (string.IsNullOrEmpty(msg)) continue;

                    queue.Enqueue(msg);
                    OnMessage?.Invoke(msg);
                }
                catch (OperationCanceledException) { return; }
                catch (ObjectDisposedException)    { return; }
                catch (Exception ex) { OnError?.Invoke(ex); }
            }
        }

        public bool TryDequeue(out string message) => queue.TryDequeue(out message!);

        public void Dispose()
        {
            if (disposed) return;
            disposed = true;
            cts.Cancel();
            udp?.Close();
            udp?.Dispose();
            cts.Dispose();
        }
    }

    /// <summary>UDP sender — send-only, no local port binding required.</summary>
    public class EdgeLinkUdpSender : IDisposable
    {
        private readonly UdpClient udp = new();
        private bool disposed;

        public Task SendAsync(string host, int port, string message)
        {
            if (disposed) throw new ObjectDisposedException(nameof(EdgeLinkUdpSender));
            byte[] bytes = Encoding.UTF8.GetBytes(message);
            return udp.SendAsync(bytes, bytes.Length, host, port);
        }

        public Task SendAsync(IPEndPoint endpoint, string message)
        {
            if (disposed) throw new ObjectDisposedException(nameof(EdgeLinkUdpSender));
            byte[] bytes = Encoding.UTF8.GetBytes(message);
            return udp.SendAsync(bytes, bytes.Length, endpoint);
        }

        public void Dispose()
        {
            if (disposed) return;
            disposed = true;
            udp.Dispose();
        }
    }
}

File: SDK/JavaScript/src/EdgeLinkUdpClient.js
"use strict";

const dgram        = require("dgram");
const EventEmitter = require("events");

/**
 * UDP receiver — binds to a local port and emits "message" for each packet.
 *
 * Events:
 *   "message" (string)
 *   "error"   (Error)
 */
class EdgeLinkUdpClient extends EventEmitter {
    /**
     * @param {number} localPort
     */
    constructor(localPort) {
        super();
        this.localPort = localPort;
        this._socket   = null;
        this.isRunning = false;
    }

    start() {
        this._socket = dgram.createSocket("udp4");

        this._socket.on("message", (buf) => {
            const msg = buf.toString("utf8").trim();
            if (msg) this.emit("message", msg);
        });

        this._socket.on("error", (err) => this.emit("error", err));

        this._socket.bind(this.localPort, () => {
            this.isRunning = true;
        });
    }

    stop() {
        if (this._socket) {
            this._socket.close();
            this._socket = null;
        }
        this.isRunning = false;
    }
}

/**
 * UDP sender — send-only, no local port binding required.
 */
class EdgeLinkUdpSender {
    constructor() {
        this._socket = dgram.createSocket("udp4");
    }

    /**
     * @param {string} host
     * @param {number} port
     * @param {string} message
     * @returns {Promise<void>}
     */
    send(host, port, message) {
        return new Promise((resolve, reject) => {
            const buf = Buffer.from(message, "utf8");
            this._socket.send(buf, 0, buf.length, port, host, (err) => {
                if (err) reject(err);
                else resolve();
            });
        });
    }

    close() {
        this._socket.close();
    }
}

module.exports = { EdgeLinkUdpClient, EdgeLinkUdpSender };
